src/QMCWaveFunctions/Jastrow/J1Spin.h

```diff
@@ -0,0 +1,854 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2021 QMCPACK developers.
+//
+// File developed by: Jeremy McMinnis, jmcminis@gmail.com, University of Illinois at Urbana-Champaign
+//                  Jeongnim Kim, jeongnim.kim@gmail.com, University of Illinois at Urbana-Champaign
+//                  Mark A. Berrill, berrillma@ornl.gov, Oak Ridge National Laboratory
+//                  Shiv Upadhyay, shivnupadhyay@gmail.com, University of Pittsburgh
+//
+// File created by: Jeongnim Kim, jeongnim.kim@intel.com, Intel Corp.
+//////////////////////////////////////////////////////////////////////////////////////
+// -*- C++ -*-
+#ifndef QMCPLUSPLUS_ONEBODYSPINJASTROW_OPTIMIZED_SOA_H
+#define QMCPLUSPLUS_ONEBODYSPINJASTROW_OPTIMIZED_SOA_H
+#include "Configuration.h"
+#include "Particle/DistanceTableData.h"
+#include "ParticleBase/ParticleAttribOps.h"
+#include "QMCWaveFunctions/WaveFunctionComponent.h"
+#include "Utilities/qmc_common.h"
+#include "Utilities/IteratorUtility.h"
+#include "CPU/SIMD/aligned_allocator.hpp"
+#include "CPU/SIMD/algorithm.hpp"
+#include <map>
+#include <numeric>
+
+namespace qmcplusplus
+{
+/** @ingroup WaveFunctionComponent
+ *  @brief Specialization for one-body Jastrow function using multiple functors
+ */
+template<class FT>
+struct J1Spin : public WaveFunctionComponent
+{
+  ///alias FuncType
+  using FuncType = FT;
+  ///type of each component U, dU, d2U;
+  using valT = typename FT::real_type;
+  ///element position type
+  using posT = TinyVector<valT, OHMMS_DIM>;
+  ///use the same container
+  using DistRow  = DistanceTableData::DistRow;
+  using DisplRow = DistanceTableData::DisplRow;
+  ///table index
+  const int myTableID;
+  ///number of ions
+  int Nions;
+  ///number of electrons
+  const int Nelec;
+  /* the number of ion groups if ions in 'Ions' particleset are grouped by species. 0 otherwise.
+   * 0 Use slow code path. >= 1 use the code path with ion grouping
+   */
+  const int NumGroups;
+  // number of electron groups
+  const int NumTargetGroups;
+  ///reference to the sources (ions)
+  const ParticleSet& Ions;
+  ///reference to the target (elecs)
+  const ParticleSet& Elecs;
```

Comment:
Keeping a reference of Elecs is ricky. Jastrow object should be tided to a  particular electron ParticleSet object.

Reply:
Removed